From 6d208bcdbcf7f5f6727e668187af25508c5e5ba5 Mon Sep 17 00:00:00 2001
From: "tao.zeng" <tao.zeng@amlogic.com>
Date: Fri, 25 Jul 2014 10:43:45 +0800
Subject: [PATCH 4571/5965] PD #94884: identify battery charge status by SAR
 ADC sign bit.

For some bad adatpers with large resistor on cable, DCIN will drop to
4.5v when loading increase, this will cause voltage limit by PMU and
battery will discharging but DCIN OK signal is still active. So we can't
identify battery status only by DCIN OK, we need SAR ADC sign bit.

Change-Id: If5c0698d5701be7030d761d0f258a5db5b7037b7
---
 drivers/amlogic/power/aml_pmu/aml1218-sply.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/drivers/amlogic/power/aml_pmu/aml1218-sply.c b/drivers/amlogic/power/aml_pmu/aml1218-sply.c
index 0d5402179259..99bb7a402b4a 100755
--- a/drivers/amlogic/power/aml_pmu/aml1218-sply.c
+++ b/drivers/amlogic/power/aml_pmu/aml1218-sply.c
@@ -1314,8 +1314,10 @@ static int aml1218_update_state(struct aml_charger *charger)
 
     charger->ibat = aml1218_get_battery_current();
     if (val & 0x18) {
-        if (charger->ibat >= 20) {
+        if (charger->ibat >= 20 && adc_sign_bit) {                      // current sign bit 1 <- charging
             charger->charge_status = CHARGER_CHARGING;                  // charging
+        } else if (!adc_sign_bit) {
+            charger->charge_status = CHARGER_DISCHARGING;               // Not charging 
         } else {
             charger->charge_status = CHARGER_NONE;                      // Not charging 
         } 
-- 
2.19.0

